File: services/rbac-service/internal/store/outbox.go
package store

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/jackc/pgx/v5"

	"github.com/windrose-ai/rbac-service/internal/events"
)

// InsertOutbox appends an event to the transactional outbox inside the
// caller's mutation transaction (MASTER-FR-034: never emit before commit).
func InsertOutbox(ctx context.Context, tx pgx.Tx, env events.Envelope) error {
	payload, err := json.Marshal(env.Payload)
	if err != nil {
		return fmt.Errorf("outbox payload: %w", err)
	}
	var viaAgent []byte
	if env.ViaAgent != nil {
		viaAgent, err = json.Marshal(env.ViaAgent)
		if err != nil {
			return fmt.Errorf("outbox via_agent: %w", err)
		}
	}
	_, err = tx.Exec(ctx, `
		INSERT INTO outbox (event_id, tenant_id, event_type, actor_type, actor_id, via_agent, resource_urn, occurred_at, trace_id, payload)
		VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)`,
		env.EventID, env.TenantID, env.EventType, env.Actor.Type, env.Actor.ID,
		viaAgent, env.ResourceURN, env.OccurredAt, env.TraceID, payload)
	if err != nil {
		return fmt.Errorf("outbox insert: %w", err)
	}
	return nil
}

// OutboxRow is one unpublished outbox entry.
type OutboxRow struct {
	ID       int64
	Envelope events.Envelope
}

// FetchUnpublished claims up to limit unpublished outbox rows in id order.
func (s *Store) FetchUnpublished(ctx context.Context, limit int) ([]OutboxRow, error) {
	var rows []OutboxRow
	err := s.WithWorker(ctx, func(tx pgx.Tx) error {
		r, err := tx.Query(ctx, `
			SELECT id, event_id, tenant_id, event_type, actor_type, actor_id, via_agent, resource_urn, occurred_at, trace_id, payload
			FROM outbox WHERE published_at IS NULL ORDER BY id LIMIT $1`, limit)
		if err != nil {
			return err
		}
		defer r.Close()
		for r.Next() {
			var row OutboxRow
			var viaAgent, payload []byte
			if err := r.Scan(&row.ID, &row.Envelope.EventID, &row.Envelope.TenantID, &row.Envelope.EventType,
				&row.Envelope.Actor.Type, &row.Envelope.Actor.ID, &viaAgent, &row.Envelope.ResourceURN,
				&row.Envelope.OccurredAt, &row.Envelope.TraceID, &payload); err != nil {
				return err
			}
			if len(viaAgent) > 0 {
				var va events.ViaAgent
				if err := json.Unmarshal(viaAgent, &va); err == nil {
					row.Envelope.ViaAgent = &va
				}
			}
			if len(payload) > 0 {
				_ = json.Unmarshal(payload, &row.Envelope.Payload)
			}
			rows = append(rows, row)
		}
		return r.Err()
	})
	return rows, err
}

// MarkPublished stamps outbox rows as published.
func (s *Store) MarkPublished(ctx context.Context, ids []int64) error {
	if len(ids) == 0 {
		return nil
	}
	return s.WithWorker(ctx, func(tx pgx.Tx) error {
		_, err := tx.Exec(ctx, `UPDATE outbox SET published_at = $1 WHERE id = ANY($2)`, time.Now().UTC(), ids)
		return err
	})
}

// OutboxEventsByType returns committed outbox envelopes of a type for a
// tenant — used by tests and the audit trail (override events, etc.).
func (s *Store) OutboxEventsByType(ctx context.Context, tenant string, eventType string) ([]events.Envelope, error) {
	var out []events.Envelope
	err := s.WithWorker(ctx, func(tx pgx.Tx) error {
		r, err := tx.Query(ctx, `
			SELECT event_id, tenant_id, event_type, actor_type, actor_id, resource_urn, occurred_at, trace_id, payload
			FROM outbox WHERE tenant_id = $1 AND event_type = $2 ORDER BY id`, tenant, eventType)
		if err != nil {
			return err
		}
		defer r.Close()
		for r.Next() {
			var env events.Envelope
			var payload []byte
			if err := r.Scan(&env.EventID, &env.TenantID, &env.EventType, &env.Actor.Type, &env.Actor.ID,
				&env.ResourceURN, &env.OccurredAt, &env.TraceID, &payload); err != nil {
				return err
			}
			if len(payload) > 0 {
				_ = json.Unmarshal(payload, &env.Payload)
			}
			out = append(out, env)
		}
		return r.Err()
	})
	return out, err
}

File: services/ingestion-service/app/events/outbox.py
"""Transactional outbox (MASTER-FR-030/031/034/035).

`emit_event` adds an outbox row to the CURRENT session/transaction — events
are never emitted before commit. `publish_pending` is the poller half
(Debezium replaces it in prod); InMemoryEventPublisher backs tests and
KafkaEventPublisher is the production stub.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any, Protocol

import sqlalchemy as sa
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.context import current_trace_id
from app.ids import uuid7
from app.store.models import OutboxEvent

TOPIC = "ingestion.events.v1"

SERVICE_ACTOR = {"type": "service", "id": "ingestion-service"}


def emit_event(
    session: AsyncSession,
    *,
    tenant_id: str,
    event_type: str,
    resource_urn: str,
    payload: dict[str, Any],
    actor: dict[str, Any] | None = None,
    via_agent: dict[str, Any] | None = None,
) -> OutboxEvent:
    event = OutboxEvent(
        id=uuid7(),
        tenant_id=tenant_id,
        event_id=uuid7(),
        event_type=event_type,
        resource_urn=resource_urn,
        actor=actor or SERVICE_ACTOR,
        via_agent=via_agent,
        occurred_at=datetime.now(UTC),
        trace_id=current_trace_id(),
        payload=payload,
    )
    session.add(event)
    return event


def envelope(event: OutboxEvent) -> dict[str, Any]:
    """MASTER-FR-031 envelope; partition key = tenant_id."""
    return {
        "event_id": event.event_id,
        "event_type": event.event_type,
        "tenant_id": event.tenant_id,
        "actor": event.actor,
        "via_agent": event.via_agent,
        "resource_urn": event.resource_urn,
        "occurred_at": event.occurred_at.isoformat(),
        "trace_id": event.trace_id,
        "payload": event.payload,
    }


class EventPublisher(Protocol):
    async def publish(self, topic: str, key: str, value: dict[str, Any]) -> None: ...


class InMemoryEventPublisher:
    def __init__(self) -> None:
        self.published: list[tuple[str, str, dict[str, Any]]] = []

    async def publish(self, topic: str, key: str, value: dict[str, Any]) -> None:
        self.published.append((topic, key, value))


class KafkaEventPublisher:
    """Real Kafka (Redpanda) publisher via the shared ``windrose_common``
    idempotent aiokafka producer. Keyed by tenant_id so a tenant's events keep a
    single-partition order (MASTER-FR-031). Runtime event publisher; the outbox
    relay drives it from committed rows."""

    def __init__(self, bootstrap_servers: str = "localhost:9092") -> None:
        from windrose_common.kafka import KafkaConfig, KafkaProducerClient

        self.bootstrap_servers = bootstrap_servers
        self._client = KafkaProducerClient(KafkaConfig(bootstrap_servers=bootstrap_servers))
        self._started = False

    async def _ensure_started(self) -> None:
        if not self._started:
            await self._client.start()
            self._started = True

    async def publish(self, topic: str, key: str, value: dict[str, Any]) -> None:
        await self._ensure_started()
        await self._client.send(topic, key, value)

    async def aclose(self) -> None:
        if self._started:
            await self._client.stop()
            self._started = False


async def publish_pending(
    session: AsyncSession, publisher: EventPublisher, limit: int = 100
) -> int:
    """Poller: publish committed outbox rows oldest-first, mark published_at.

    Drains across every tenant, so it cannot rely on a per-request
    `tenant_session` (app.tenant_id GUC). On Postgres, the runtime role is
    NOSUPERUSER NOBYPASSRLS (migration 0004), so a plain SELECT would hit the
    `tenant_isolation` policy with no GUC set. Instead this claims/marks rows
    through the `ing_outbox_claim_pending`/`ing_outbox_mark_published`
    SECURITY DEFINER functions (migration 0005) -- an RLS bypass scoped to
    just the outbox table. SQLite (unit tier, no RLS) uses a plain query.
    """
    dialect = session.get_bind().dialect.name
    if dialect == "postgresql":
        rows = (
            (
                await session.execute(
                    sa.select(OutboxEvent).from_statement(
                        sa.text("SELECT * FROM ing_outbox_claim_pending(:lim)")
                    ),
                    {"lim": limit},
                )
            )
            .scalars()
            .all()
        )
    else:
        rows = (
            (
                await session.execute(
                    sa.select(OutboxEvent)
                    .where(OutboxEvent.published_at.is_(None))
                    .order_by(OutboxEvent.occurred_at)
                    .limit(limit)
                )
            )
            .scalars()
            .all()
        )
    if not rows:
        return 0
    for event in rows:
        await publisher.publish(TOPIC, event.tenant_id, envelope(event))
    if dialect == "postgresql":
        await session.execute(
            sa.text("SELECT ing_outbox_mark_published(CAST(:ids AS uuid[]))"),
            {"ids": [event.id for event in rows]},
        )
    else:
        for event in rows:
            event.published_at = datetime.now(UTC)
    await session.commit()
    return len(rows)
